sql/atxs/atxs.go

```diff
@@ -0,0 +1,213 @@
+package atxs
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/sql"
+)
+
+// GetATXByID gets an ATX by a given ATX ID.
+func GetATXByID(db sql.Executor, id types.ATXID) (atx *types.ActivationTx, err error) {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindBytes(1, id.Bytes())
+	}
+	dec := func(stmt *sql.Statement) bool {
+		var v types.ActivationTx
+		if _, decodeErr := codec.DecodeFrom(stmt.ColumnReader(0), &v); decodeErr != nil {
+			atx, err = nil, fmt.Errorf("decode %w", decodeErr)
+			return true
+		}
+
+		v.SetID(&id)
+		atx, err = &v, nil
+		return true
+	}
+
+	if rows, err := db.Exec("select atx from atxs where id = ?1;", enc, dec); err != nil {
+		return nil, fmt.Errorf("exec id %v: %w", id, err)
+	} else if rows == 0 {
+		return nil, fmt.Errorf("exec id %v: %w", id, sql.ErrNotFound)
+	}
+
+	return atx, err
+}
+
+// HasID checks if an ATX exists by a given ATX ID.
+func HasID(db sql.Executor, id types.ATXID) (bool, error) {
+	rows, err := db.Exec("select 1 from atxs where id = ?1;",
+		func(stmt *sql.Statement) {
+			stmt.BindBytes(1, id.Bytes())
+		}, nil,
+	)
+	if err != nil {
+		return false, fmt.Errorf("exec id %v: %w", id, err)
+	}
+	return rows > 0, nil
+}
+
+// GetTimestampByID gets an ATX timestamp by a given ATX ID.
+func GetTimestampByID(db sql.Executor, id types.ATXID) (timestamp time.Time, err error) {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindBytes(1, id.Bytes())
+	}
+	dec := func(stmt *sql.Statement) bool {
+		timestamp = time.Unix(0, stmt.ColumnInt64(0))
+		return true
+	}
+
+	if rows, err := db.Exec("select timestamp from atxs where id = ?1;", enc, dec); err != nil {
+		return time.Time{}, fmt.Errorf("exec id %v: %w", id, err)
+	} else if rows == 0 {
+		return time.Time{}, fmt.Errorf("exec id %s: %w", id, sql.ErrNotFound)
+	}
+
+	return timestamp, err
+}
+
+// GetLastIDByNodeID gets the last ATX ID for a given node ID.
+func GetLastIDByNodeID(db sql.Executor, nodeID types.NodeID) (id types.ATXID, err error) {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindBytes(1, nodeID.ToBytes())
+	}
+	dec := func(stmt *sql.Statement) bool {
+		stmt.ColumnBytes(0, id[:])
+		return true
+	}
+
+	if rows, err := db.Exec(`
+		select id from atxs 
+		where smesher = ?1
+		order by timestamp desc 
+		limit 1;`, enc, dec); err != nil {
+		return *types.EmptyATXID, fmt.Errorf("exec id %v: %w", id, err)
```

Comment:
this approach is very strange, can we just initialize empty atx id like types.ATXID{}